`make test` failed with below error:

```
TASK: [sansible.gocd_server | Install go-server] ****************************** 

failed: [default] => {"failed": true}
stderr: W: --force-yes is deprecated, use one of the options starting with --allow instead.
E: Failed to fetch https://download.go.cd/binaries/16.1.0-2855/deb/go-server-16.1.0-2855.deb  transfer closed with 119295513 bytes remaining to read

E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?

stdout: Reading package lists...
Building dependency tree...
Reading state information...
Suggested packages:
  go-agent
The following NEW packages will be installed:
  go-server
0 upgraded, 1 newly installed, 0 to remove and 24 not upgraded.
Need to get 141 MB of archives.
After this operation, 419 MB of additional disk space will be used.
Get:1 https://download.go.cd  go-server 16.1.0-2855 [141 MB]
Err:1 https://download.go.cd  go-server 16.1.0-2855
  transfer closed with 119295513 bytes remaining to read

msg: '/usr/bin/apt-get -y -o "Dpkg::Options::=--force-confdef" -o "Dpkg::Options::=--force-confold" --force-yes  install 'go-server=16.1.*'' failed: W: --force-yes is deprecated, use one of the options starting with --allow instead.
E: Failed to fetch https://download.go.cd/binaries/16.1.0-2855/deb/go-server-16.1.0-2855.deb  transfer closed with 119295513 bytes remaining to read

E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?


FATAL: all hosts have already failed -- aborting

PLAY RECAP ******************************************************************** 
           to retry, use: --limit @/home/hkuehn/test.retry

default                    : ok=17   changed=13   unreachable=0    failed=1   

Ansible failed to complete successfully. Any error output should be
visible above. Please fix these errors and try again.
Makefile:25: recipe for target 'vagrant' failed
make: *** [vagrant] Error 1
```

Ubuntu 16.04 requires more fine grained `apt` options. Alternative options are `--allow-downgrades`,`--allow-remove-essential` ,`--allow-change-held-packages`.

http://manpages.ubuntu.com/manpages/xenial/man8/apt-get.8.html

--- a/.gitignore
+++ b/.gitignore
@@ -1,5 +1,6 @@
 .vagrant
 /meta/.galaxy_install_info
-/tests/vagrant/sansible.*
-/tests/vagrant/local_vars.yml
+/tests/sansible.*
 *.bak
+*.retry
+*.log

--- a/.travis.yml
+++ b/.travis.yml
@@ -28,7 +28,7 @@ script:
   # Install Role Dependencies and lint
   - make test_deps lint 
   # First run
-  - cd tests/vagrant/
+  - cd tests/
   # Test the "build" stage
   - ansible-playbook -i inventory test.yml --tags=build --connection=local
   - >
@@ -43,6 +43,8 @@ script:
     | grep -q 'changed=0.*failed=0'
     && (echo 'Idempotence test: pass' && exit 0)
     || (echo 'Idempotence test: fail' && exit 1)
+  # Test the "assert" stage
+  - ansible-playbook -i inventory test.yml --tags=assert --connection=local
     
 notifications:
   webhooks: https://galaxy.ansible.com/api/v1/notifications/

--- a/Makefile
+++ b/Makefile
@@ -4,23 +4,21 @@
 
 all: test vagrant_halt clean
 
-all: test clean
-
 ## Run tests on any file change
 watch: test_deps
 	while sleep 1; do \
-		find defaults/ handlers/ meta/ tasks/ templates/ tests/vagrant/test.yml \
-		| entr -d make test; \
+		find defaults/ meta/ tasks/ templates/ tests/test.yml tests/vagrant/Vagrantfile \
+		| entr -d make lint vagrant; \
 	done
 
 ## Run tests
-test: test_deps vagrant_up
+test: test_deps lint vagrant
 
 ## Install test dependencies
 test_deps:
-	rm -rf tests/vagrant/sansible.gocd_server
-	ln -s ../.. tests/vagrant/sansible.gocd_server
-	ansible-galaxy install --force -p tests/vagrant -r tests/vagrant/local_requirements.yml
+	rm -rf tests/sansible.*
+	ln -s .. tests/sansible.gocd_server
+	ansible-galaxy install --force -p tests/ -r tests/local_requirements.yml
 
 ## Start and (re)provisiom Vagrant test box
 vagrant:
@@ -40,7 +38,7 @@ lint:
 
 ## Clean up
 clean:
-	rm -rf tests/vagrant/sansible.*
+	rm -rf tests/sansible.*
 	cd tests/vagrant && vagrant destroy
 
 ## Prints this help

--- a/README.md
+++ b/README.md
@@ -1,5 +1,8 @@
 # GO CD Server
 
+Please see the Sansible readme for information on how to 
+[contribure](https://github.com/sansible/sansible)
+
 Master: [![Build Status](https://travis-ci.org/sansible/gocd_server.svg?branch=master)](https://travis-ci.org/sansible/gocd_server)  
 Develop: [![Build Status](https://travis-ci.org/sansible/gocd_server.svg?branch=develop)](https://travis-ci.org/sansible/gocd_server)
 
@@ -9,7 +12,8 @@ Develop: [![Build Status](https://travis-ci.org/sansible/gocd_server.svg?branch=
 * [Examples](#examples)
 * [Development & Testing](#development---testing)
 
-This roles installs GO CD server.
+This roles installs GO CD server, for GO CD Agent installation please 
+see the [Sansible GO CD Agent Role](https://github.com/sansible/gocd_agent)
 
 For more information about GO CD please visit [go.cd/](http://www.go.cd/).
 
@@ -179,20 +183,25 @@ create ~/.aws/credentials for you. But we strongly advise to use
 
 ## Development & Testing
 
-If you want to work on this role, please start with running
-`make watch`. This will watch for any file changes and re-provision
-vagrant box.
+If you want to work on this role you will find several test targets 
+in the Makfefile.
 
-If you want to test the AWS backup/restore script locally you need
-to create a file `/tests/vagrant/local_vars.yml` with
+`make test` - This will bring up the Vagrant machine and run a test
+at the end  
+`make watch` - This will watch for file system changes and 
+reprovision the Vagrant machine.
 
-```YAML
----
+You will need to have ansible-lint installed to run the tests:
 
-gocd_server:
-  aws:
-    backup_bucket: s3://backup.gocd-server.my.domain
-    access_key_id: LOREMIPSUMDOLORKV4IQ
-    secret_access_key: sit+Ament+Hl8V2lQpaOkLRakvGMidrkWv6F9
+```BASH
+pip install ansible-lint
+```
 
+By default the tests will bring an Ubuntu Trust64 machine, you can make them
+bring up a Xenial 64 machine by adding ```UBUNTU_VERSION=16``` before 
+these make commands, eg:
+
+```BASH
+UBUNTU_VERSION=16 make test
+UBUNTU_VERSION=16 make watch
 ```

--- a/defaults/main.yml
+++ b/defaults/main.yml
@@ -13,7 +13,7 @@ gocd_server:
   server_min_perm_gen: 128m
   start_opts: ""
   user: go
-  version: 16.1.*
+  version: 16.6.*
 
   packages:
     - git

--- a/meta/main.yml
+++ b/meta/main.yml
@@ -10,13 +10,14 @@ galaxy_info:
       versions:
         - precise
         - trusty
+        - xenial
   categories:
     - development
     - cd
 
 dependencies:
   - role: sansible.users_and_groups
-    version: v1.0
+    version: v1.2
     users_and_groups:
       users:
         - name: "{{ gocd_server.user }}"

--- a/tests/ansible.cfg
+++ b/tests/ansible.cfg
@@ -1,0 +1,3 @@
+[defaults]
+allow_world_readable_tmpfiles=True
+hash_behaviour = merge

--- a/tests/inventory
+++ b/tests/inventory
@@ -1,0 +1,1 @@
+localhost

--- a/tests/local_requirements.yml
+++ b/tests/local_requirements.yml
@@ -1,0 +1,7 @@
+---
+
+- src: sansible.java
+  version: v1.0
+
+- src: sansible.users_and_groups
+  version: v1.2

--- a/tests/test.yml
+++ b/tests/test.yml
@@ -1,0 +1,28 @@
+---
+
+- name: Test GOCD Server
+  hosts: all
+
+  pre_tasks:
+    - name: Update apt
+      become: yes
+      apt:
+        cache_valid_time: 1800
+        update_cache: yes
+      tags:
+        - build
+
+  roles:
+    - role: sansible.gocd_server
+
+  post_tasks:
+    - name: GoCD Server should already be started
+      become: yes
+      service:
+        name: go-server
+        state: started
+      register: server_start_check
+      failed_when: server_start_check.changed
+      tags:
+        - assert
+

--- a/tests/vagrant/Vagrantfile
+++ b/tests/vagrant/Vagrantfile
@@ -1,29 +1,20 @@
 # -*- mode: ruby -*-
 # vi: set ft=ruby :
 
-require 'yaml'
-local_vars = {
-	'vagrant' => {
-		'priv_ip' => nil
-	}
-}.merge(
-	File.file?('local_vars.yml') ?
-	YAML.load_file('local_vars.yml') : {}
-)
-
 # Vagrantfile API/syntax version. Don't touch unless you know what you're doing!
 VAGRANTFILE_API_VERSION = "2"
 
 Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
 
-	config.vm.box = "ubuntu/trusty64"
+	if ENV['UBUNTU_VERSION'] == "16"
+		config.vm.box = "ubuntu/xenial64"
+	else
+		config.vm.box = "ubuntu/trusty64"
+	end
+
 	config.vm.box_check_update = false
 
 	config.ssh.forward_agent = true
-
-	if local_vars['vagrant']['priv_ip']
-		config.vm.network :private_network, ip: local_vars['vagrant']['priv_ip']
-	end
 
 	config.vm.provider :virtualbox do |vb|
 		vb.customize ["modifyvm", :id, "--memory", 1024]
@@ -33,19 +24,21 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
 		vb.customize ["modifyvm", :id, "--natdnsproxy1", "on"]
 	end
 
+	if ENV['UBUNTU_VERSION'] == "16"
+		config.vm.provision "shell", inline: "apt-get install -y python-minimal"
+	end
+
 	# Test "build" tag
 	config.vm.provision "ansible" do |ansible|
-		ansible.playbook = "test.yml"
+		ansible.playbook = "../test.yml"
 		ansible.tags = [ "build" ]
-		ansible.extra_vars = local_vars
 		ansible.verbose = "#{ENV['VERBOSE']}" if ENV['VERBOSE']
 	end
 
 	# Test "configure" tag
 	config.vm.provision "ansible" do |ansible|
-		ansible.playbook = "test.yml"
-		ansible.tags = [ "configure" ]
-		ansible.extra_vars = local_vars
+		ansible.playbook = "../test.yml"
+		ansible.tags = [ "configure", "assert" ]
 		ansible.verbose = "#{ENV['VERBOSE']}" if ENV['VERBOSE']
 	end
 end

--- a/tests/vagrant/ansible.cfg
+++ b/tests/vagrant/ansible.cfg
@@ -1,2 +1,1 @@
-[defaults]
-hash_behaviour = merge
+../ansible.cfg

